### PR TITLE
feat: gaslimit encoding end to end

### PR DESCRIPTION
**Pull request by tynes**
**Description**
This PR included end to end testing of the gasLimit encoding scheme described in https://github.com/ethereum-optimism/optimism/issues/823#issuecomment-844557937

A JS fees library was written that implements the scheme and is added to `core-utils`. It could use some more tests but currently passes the end to end tests. Will follow up with more tests as well as https://github.com/ethereum-optimism/optimism/issues/823#issuecomment-845554638

It has some additional implementation details, for example handling gasPrices set to 0

### CODE REVIEW

File: packages/core-utils/src/fees.ts [hunk 1]
```diff
@@ -0,0 +1,95 @@
...
+const txDataNonZeroGasEIP2028 = 16
+const overhead = 4200
+
+function encode(
+  data: Buffer | string,
+  l1GasPrice: BigNumber,
+  l2GasLimit: BigNumber,
+  l2GasPrice: BigNumber
+): BigNumber {
+  const l1GasLimit = calculateL1GasLimit(data)
```
Comment (tynes) [inline packages/core-utils/src/fees.ts:19]: This should throw an error if the values do not satisfy the correct equations, see https://github.com/ethereum-optimism/optimism/issues/823#issuecomment-844557937

File: integration-tests/test/rpc.spec.ts [hunk 2]
```diff
@@ -297,28 +297,34 @@ describe('Basic RPC tests', () => {
...
-        // we normalize by gwei here because the RPC does it as well, since the
-        // user provides a 1gwei gas price when submitting txs via the eth_gasPrice
-        // rpc call. The smallest possible value for the expected cost is 21000
-        let expected = expectedCost[i].mul(l1GasPrice).div(GWEI)
-        if (expected.lt(BigNumber.from(21000))) {
-          expected = BigNumber.from(21000)
         }
-        expect(estimate).to.be.deep.eq(expected)
+        const estimate = await l2Provider.estimateGas(tx)
+        const l2Gaslimit = await l2Provider.send('eth_estimateExecutionGas', [
```
Comment (tynes) [inline integration-tests/test/rpc.spec.ts:310]: This new RPC endpoint was added to be exactly what the L1 `eth_estimateGas` is

File: packages/core-utils/test/fees.spec.ts [hunk 3]
```diff
@@ -0,0 +1,19 @@
+import './setup'
```
Comment (tynes) [inline packages/core-utils/test/fees.spec.ts:1]: This needs more tests but the e2e tests are passing using `encode` and `decode`

**Comment by tynes** [comment]
This implementation needs to use `big.Int`s for handling gas prices in geth

File: l2geth/cmd/utils/flags.go [hunk 4]
```diff
@@ -894,12 +894,17 @@ var (
...
 	RollupEnableL2GasPollingFlag = cli.BoolFlag{
 		Name:   "rollup.enablel2gaspolling",
-		Usage:  "",
+		Usage:  "Poll for the L2 gas price from the L2 state",
+		EnvVar: "ROLLUP_ENABLE_L2_GAS_POLLING",
+	}
+	RollupEnforceFeesFlag = cli.BoolFlag{
+		Name:   "rollup.enforcefeesflag",
+		Usage:  "Disable transactions with 0 gas price",
 		EnvVar: "ROLLUP_ENABLE_L2_GAS_POLLING",
```
Comment (tynes) [inline l2geth/cmd/utils/flags.go:908]: Fix this env var

File: l2geth/core/rollup_fee.go [hunk 5]
```diff
@@ -64,15 +67,15 @@ func RoundL1GasPrice(gasPrice uint64) uint64 {
```
Comment (tynes) [inline l2geth/core/rollup_fee.go:67]:
```
func RoundL1GasPrice(gasPrice uint64) uint64 {
	if gasPrice%hundredMillion == 0 {
		return gasPrice
	}
	gasPrice += hundredMillion
	return gasPrice - gasPrice%hundredMillion
}
```

Comment (karlfloersch) [reply l2geth/core/rollup_fee.go:67]:
```
func CeilModOneHundredMillion(num uint64) uint64 {
	if num%hundredMillion == 0 {
		return num
	}
	num += hundredMillion
	return num - num%hundredMillion
}

func RoundL1GasPrice(gasPrice uint64) uint64 {
  return CeilModOneHundredMillion(gasPrice)
}

func RoundL2GasPrice(gasPrice uint64) uint64 {
  if gasPrice == 0 {
    return 1
  }
  if gasPrice == 1 {
    return hundredMillion + 1
  }
  return CeilModOneHundredMillion(gasPrice - 1) + 1
}

```

File: l2geth/rollup/sync_service.go [hunk 6]
```diff
@@ -93,6 +94,7 @@ func NewSyncService(ctx context.Context, cfg Config, txpool *core.TxPool, bc *co
 	// Initialize the rollup client
 	client := NewClient(cfg.RollupClientHttp, chainID)
 	log.Info("Configured rollup client", "url", cfg.RollupClientHttp, "chain-id", chainID.Uint64(), "ctc-deploy-height", cfg.CanonicalTransactionChainDeployHeight)
+	log.Info("Enforce Fees", cfg.EnforceFees)
```
Comment (tynes) [inline l2geth/rollup/sync_service.go:97]: Need another string arg

File: l2geth/rollup/sync_service.go [hunk 7]
```diff
@@ -749,6 +754,33 @@ func (s *SyncService) applyTransaction(tx *types.Transaction) error {
...
+	}
+
+	gasUsed := core.DecodeL2GasLimit(tx.Gas())
+	l2GasLimit := new(big.Int).SetUint64(gasUsed)
+	fee, err := core.CalculateRollupFee(tx.Data(), l1GasPrice, l2GasLimit, l2GasPrice)
+	if err != nil {
+		return err
+	}
+
+	if !s.enforceFees && tx.GasPrice().Cmp(common.Big0) == 0 {
```
Comment (tynes) [inline l2geth/rollup/sync_service.go:774]: This should be able to be moved to the top

**Comment by tynes** [comment]
I have more commits coming for this branch soon

File: l2geth/core/rollup_fee.go [hunk 8]
```diff
@@ -47,34 +54,58 @@ func calculateL1GasLimit(data []byte, overhead uint64) *big.Int {
...
+		return errInvalidGasPrice
 	}
-	return gasPrice - gasPrice%hundredMillion
+	return nil
 }
 
-func RoundL2GasPrice(gasPrice uint64) uint64 {
-	if gasPrice == 0 {
-		return gasPrice
+// VerifyL2GasPrice returns an error if the number is an invalid possible L2 gas
```
Comment (tynes) [inline l2geth/core/rollup_fee.go:77]: `big.Int`s are too verbose but we have to use them because we need the arbitrary precision

**Pull request merged**